doc/source/development.rst

```diff
@@ -1,80 +1,88 @@
-   often useful to start them in a debugger (``gdb`` on Linux or ``lldb`` on
-   MacOS). See the latest discussion about how to do this `here`_.
-
+   directory will take effect without reinstalling the package. In
+   contrast, if you do ``python setup.py install``, files will be copied
+   from the Ray directory to a directory of Python packages (often
+   something like
+   ``/home/ubuntu/anaconda3/lib/python3.6/site-packages/ray``). This
+   means that changes you make to files in the Ray directory will not
+   have any effect.
+
+   If you run into **Permission Denied** errors when running
+   ``pip install``, you can try adding ``--user``. You may also need to
+   run something like ``sudo chown -R $USER /home/ubuntu/anaconda3``
+   (substituting in the appropriate path).
+
+   If you make changes to the C++ files, you will need to recompile
+   them. However, you do not need to rerun ``pip install -e .``.
+   Instead, you can recompile much more quickly by doing
+
+   .. code:: shell
+
+      cd ray/python/ray/core
+      make -j8
+
+2. **Starting processes in a debugger:** When processes are crashing, it
+   is often useful to start them in a debugger (``gdb`` on Linux or
+   ``lldb`` on MacOS). See the latest discussion about how to do this
+   here https://github.com/ray-project/ray/issues/108.
 3. **Running tests locally:** Suppose that one of the tests (e.g.,
    ``runtest.py``) is failing. You can run that test locally by running
-   ``python test/runtest.py``. However, doing so will run all of the tests which
-   can take a while. To run a specific test that is failing, you can do
+   ``python test/runtest.py``. However, doing so will run all of the
+   tests which can take a while. To run a specific test that is failing,
+   you can do
 
-   .. code-block:: shell
+   .. code:: shell
 
-     cd ray
-     python test/runtest.py APITest.testKeywordArgs
+      cd ray
+      python test/runtest.py APITest.testKeywordArgs
 
-   When running tests, usually only the first test failure matters. A single
-   test failure often triggers the failure of subsequent tests in the same
-   script.
+   When running tests, usually only the first test failure matters. A
+   single test failure often triggers the failure of subsequent tests in
+   the same script.
 
-4. **Running linter locally:** To run the Python linter on a specific file, run
-   something like ``flake8 ray/python/ray/worker.py``. You may need to first run
-   ``pip install flake8``.
+4. **Running linter locally:** To run the Python linter on a specific
+   file, run something like ``flake8 ray/python/ray/worker.py``. You may
+   need to first run ``pip install flake8``.
+5. **Autoformatting code**. We use ``yapf``
+   https://github.com/google/yapf for linting, and the config file is
+   located at ``.style.yapf``. We recommend adding ``.travis/yapf.sh``
+   to git's ``pre-push`` hook, so that you never have to worry about
```

Comment:
This is too slow to run on each push. How do you execute yapf against a subset of files, or only changed files?